### ISSUE TITLE
The logic that calculates the recommended release of a project doesn't work as expected.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5833]
This issue is possibly related to others, like #2698 and #3128.

I am opening this specifically for the use case where a new major release is created by the project maintainer, however the older major version is still shown as the recommended.

See most recent thread in Zulip: https://backdrop.zulipchat.com/#narrow/stream/218635-Backdrop/topic/Change.20of.20.22Recommended.20release.22.20for.20Comments.20Block

@argiepiano 
> Could someone please update the b-org module portal to make version 1.x-3.0.0 the "Recommended version" for the module Comments Block? Currently version 1.x-2.5.1 is shown as the recommended one.

How we fix this every time someone in the community reports this problem, is to go the project node for the specific project in b.org, and then on the "Releases" tab select the major release that we want to make the recommended one from the "Recommended major version" select:

![image](https://user-images.githubusercontent.com/2423362/195794551-35c9bd29-a4cd-4ddb-b1df-f452dfabfb8a.png)

This should not be happening, since the purpose of that select is to allow forcing a previous major version as the recommended, in exceptional cases where there is a valid reason for this - it is NOT there to be manually selecting the recommended version every time there's a new major release. In other words, that select should always be set to the default "None", and things should work as expected without any need to manually intervene - it is there only to allow overriding the recommended version.

In `update_calculate_project_update_status()`, we have the following explanation in the docblock:

```php
...
 * Given a target major version, the available releases are scanned looking for
 * the specific release to recommend (avoiding beta releases and development
 * snapshots if possible). For the target major version, the highest patch level
 * is found. If there is a release at that patch level with no extra ("beta",
 * etc.), then the release at that patch level with the most recent release date
 * is recommended. If every release at that patch level has extra (only betas),
 * then the latest release from the previous patch level is recommended. For
 * example:
 *
 * - 1.6-bugfix <-- recommended version because 1.6 already exists.
 * - 1.6
 *
 * or
 *
 * - 1.6-beta
 * - 1.5 <-- recommended version because no 1.6 exists.
 * - 1.4
 *
 * Also, the latest release from the same major version is looked for, even beta
 * releases, to display to the user as the "Latest version" option.
 * Additionally, the latest official release from any higher major versions that
 * have been released is searched for to provide a set of "Also available"
 * options.
...
```

So I think that the logic in that function is not working as expected.

**Comment by klonos** [comment]
I've reverted the "Recommended major version" for https://backdropcms.org/project/commentsblock back to the default "none", and I've run some tests on my local. The culprit seems to be what is being sent from b.org (I'm assuming the Project module) - NOT the logic in Backdrop core after all.

This is what is being fetched by `_update_process_fetch_task()`:

```xml
<?xml version="1.0" encoding="utf-8"?>
<project xmlns:dc="http://purl.org/dc/elements/1.1/">
  <title>Comments block</title>
  <short_name>commentsblock</short_name>
  <dc:creator>drop</dc:creator>
  <type>project_module</type>
  <api_version>1.x</api_version>
  <project_status>published</project_status>
  <supported_majors>2,3</supported_majors>
  <default_major>2</default_major> ← ← ← ← ← ← ← ← ← ← ←    N O T I C E    T H I S  👎🏼 
  <link>https://backdropcms.org/project/commentsblock</link>
  <releases>
...
```

Then I've set the recommended version to `3`, cleared caches, and tried again:

```xml
<?xml version="1.0" encoding="utf-8"?>
<project xmlns:dc="http://purl.org/dc/elements/1.1/">
  <title>Comments block</title>
  <short_name>commentsblock</short_name>
  <dc:creator>drop</dc:creator>
  <type>project_module</type>
  <api_version>1.x</api_version>
  <project_status>published</project_status>
  <recommended_major>3</recommended_major> ← ← ← ← ← ← ← ← ← ← ←    N O T I C E    T H I S  👍🏼 
  <supported_majors>2,3</supported_majors>
  <default_major>2</default_major> ← ← ← ← ← ← ← ← ← ← ←    A N D    T H I S  👎🏼 
  <link>https://backdropcms.org/project/commentsblock</link>
  <releases>
...
```

So my assumption here is that the Project module in b.org is only setting the `default_major` to whatever is the first major version to be created for a project, and then it fails to update that when new major releases are created.

**Comment by klonos** [comment]
Can I please have someone confirm my findings, so that we can close this issue and focus on fixing the logic that sets the `default_major` in b.org?

**Comment by indigoxela** [comment]
> Can I please have someone confirm my findings

I can with a contrib theme - the default major stays the same, when updating the recommended version.

https://updates.backdropcms.org/release-history/snazzy/1.x

Not sure, how this is supposed to be, though.

**Comment by indigoxela** [comment]
Aha... https://github.com/backdrop-contrib/project/blob/a0b5b1147ab2f1ba274d798e0424ec1511fe8406/project_release/project_release.cron.inc#L206

It is supposed to be the minimal supported version.

**Comment by klonos** [comment]
Thanks for digging that out @indigoxela 🙏🏼 

I think that we should either change that logic in project_release to be the `max()` instead of the `min()` (that makes sense to me), or then change the logic in core so to NOT take that into account as the default selected when installing new modules and/or not show that as the recommended if no recommended release has been explicitly defined.

**Comment by klonos** [comment]
> If there's NO recommended major, it falls back to the oldest major - when would that make sense?

Yeah if that's the case, then it doesn't sound right. If there's no recommended major, it should fall back to the max major with a stable (non-beta/non-alpha) release.

**Comment by indigoxela** [comment]
Quick update on that "fallback" behavior:

This would work in the project browser in B installs, but this would *not* work at all on B-org. The view doesn't show any projects without recommended release set. We discovered that recently. So this whole "recommended" thing (on B-org, in the Project contrib module, in the Project browser, in the Installer core module...) has to be handled with care.
